api/v1beta1/types_class.go

```diff
@@ -94,7 +94,7 @@ type AzureManagedControlPlaneClassSpec struct {
 
 	// VirtualNetwork describes the virtual network for the AKS cluster. It will be created if it does not already exist.
 	// +optional
-	VirtualNetwork ManagedControlPlaneVirtualNetwork `json:"virtualNetwork,omitempty"`
+	VirtualNetwork ManagedControlPlaneVirtualNetworkClassSpec `json:"virtualNetwork,omitempty"`
```

Comment:
I don't know how we can make these changes without breaking the API contract

Reply:
Yeah it's tricky. Also doing it this way messes with the CRDs per my comment above. I think I'm misunderstanding how template types work when they are a field within a template type. For example, we want some fields of `VirtualNetwork` to be shared while others (like `name`) shouldn't. This pattern exists for `FleetsMember` but doing so for the vnet doesn't seem to work https://github.com/willie-yao/cluster-api-provider-azure/blob/a3758946dbc5de3b5941100d41b599328e42a892/api/v1beta1/types_class.go#L225

I'll do some more digging into this to figure out a solution!

Reply:
If we are unable to make an api change here, would it be possible to use webhook validation to block setting `name` if the amcp is being created from a template? It's definitely not ideal.

Reply:
I would still classify making this invalid in the webhook as a breaking change I think, at least in the sense that "I have this YAML that used to work but now doesn't."

I think the most we can do here without breaking anyone is to update our examples to patch in the vnet name per-cluster vs. defining it on the template, then make the webhook return a warning if users do define it in the template to say "you probably don't want to do this."

Reply:
Made this change in this commit: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4735/commits/f3f78a757967b5d60b11931a0127f8cf45e21867

It basically overwrites the field if someone sets the vnet name from the template. Is that the right idea?

Reply:
If I'm allowed to split hairs, I would still call that a breaking behavioral change that I think still falls within the contract if a user defines the vnet name in a template and that's no longer honored by CAPZ.

Reply:
Ah okay I think I misread your comment. When you say "update our examples" do you just mean the doc examples? And the only thing we should do here is add a webhook warning.

Reply:
The docs at a minimum, but I think we should also update all the templates in order to follow those new docs. A webhook warning still sgtm.

Reply:
Gotcha. None of our templates explicitly specify the `virtualNetwork` field in ClusterClass so we're good on that front. I'll update the docs and keep the webhook warning!